ruma-serde/src/canonical_json/value.rs

```diff
@@ -0,0 +1,244 @@
+            if s.len() > 65_535 {
+                Err(serde_json::Error::custom("TOO LARGE"))
+            } else {
+                Ok(s)
+            }
+        })?)
+    }
+}
+
+impl fmt::Debug for CanonicalJsonValue {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            CanonicalJsonValue::Null => formatter.debug_tuple("Null").finish(),
+            CanonicalJsonValue::Bool(v) => formatter.debug_tuple("Bool").field(&v).finish(),
+            CanonicalJsonValue::Integer(ref v) => fmt::Debug::fmt(v, formatter),
+            CanonicalJsonValue::String(ref v) => formatter.debug_tuple("String").field(v).finish(),
+            CanonicalJsonValue::Array(ref v) => {
+                formatter.write_str("Array(")?;
+                fmt::Debug::fmt(v, formatter)?;
+                formatter.write_str(")")
+            }
+            CanonicalJsonValue::Object(ref v) => {
+                formatter.write_str("Object(")?;
+                fmt::Debug::fmt(v, formatter)?;
+                formatter.write_str(")")
+            }
+        }
+    }
+}
+
+impl fmt::Display for CanonicalJsonValue {
+    /// Display a JSON value as a string.
+    ///
+    /// ```
+    /// # use serde_json::json;
+    /// #
+    /// let json = json!({ "city": "London", "street": "10 Downing Street" });
+    ///
+    /// // Canonical format:
+    /// //
+    /// // {"city":"London","street":"10 Downing Street"}
+    /// let compact = format!("{}", json);
+    /// assert_eq!(compact,
+    ///     "{\"city\":\"London\",\"street\":\"10 Downing Street\"}");
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", serde_json::to_string(&self).map_err(|_| fmt::Error)?)
+    }
+}
+
+impl TryFrom<JsonValue> for CanonicalJsonValue {
+    type Error = serde_json::Error;
+    fn try_from(json: JsonValue) -> Result<Self, Self::Error> {
+        Ok(match json {
+            JsonValue::Bool(b) => Self::Bool(b),
+            JsonValue::Number(num) => Self::Integer(
+                Int::try_from(num.as_i64().ok_or_else(|| {
+                    serde_json::Error::custom("Invalid number found, expected i64")
+                })?)
+                .map_err(serde_json::Error::custom)?,
+            ),
+            JsonValue::Array(vec) => {
+                Self::Array(vec.into_iter().map(TryInto::try_into).collect::<Result<Vec<_>, _>>()?)
+            }
+            JsonValue::String(string) => Self::String(string),
+            JsonValue::Object(obj) => Self::Object(
+                obj.into_iter()
+                    .map(|(k, v)| Ok((k, v.try_into()?)))
+                    .collect::<Result<BTreeMap<_, _>, _>>()?,
+            ),
+            JsonValue::Null => Self::Null,
+        })
+    }
+}
+
+impl serde::ser::Serialize for CanonicalJsonValue {
+    #[inline]
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: ::serde::Serializer,
+    {
+        match *self {
```

Comment:
I think I've removed all matches with `ref` elsewhere, this can also be rewritten to use `match self` + dereferencing where necessary (only for `::Bool` I think).